text/0010-bft-signatures.md

```diff
@@ -0,0 +1,291 @@
+Accordingly, the procedure to build a signed data element from a `MetadataSignature` will first look if the `signature_header` is not empty, and if so it will proceed to regular processing.
+Otherwise, if the `signature_header` is empty, an ephemeral one will be reconstructed from the `nonce`, `signer_id` and the latest channel config and then procesing can proceed as formerly mentioned.
+
+# Implementation plan
+[implementation]: #implementation-plan
+
+The following components are to be changed:
+
+- **Channel configuration**: Addition of `ConsenterMapping` to the channel configuration (gated behind a V3.0 [capability](https://hyperledger-fabric.readthedocs.io/en/latest/capabilities_concept.html)).
+- **Peer and orderer local configuration**: Addition of BFT or CFT block replication toggle (automatic, BFT, CFT).
+- **Orderer block replication primitive**: The `BlockPuller` is an [abstraction](https://github.com/hyperledger/fabric/blob/main/orderer/consensus/etcdraft/chain.go#L83-L87) that is given to leader based consensus plugins 
+and is backed by underlying code in the `orderer/common/cluster` package. The abstraction can remain as-is, but its underlying implementation
+  would need to change in order to be able to pull attestations of block existence (blocks without transactions). 
+  Similarly,the same underlying implementation is used when onboarding orderers.
+- **Peer block replication primitive**: The peer currently contains an [implementation](https://github.com/hyperledger/fabric/blob/main/internal/pkg/peer/blocksprovider/blocksprovider.go) that fetches blocks from a single orderer.
+The implementation would need to be changed accordingly to protect against block denial of service attacks.
+- **Common deliver service**: The [common deliver service](https://github.com/hyperledger/fabric/blob/main/common/deliver/deliver.go), used by both peers and orderers when sending blocks to subscribers, will be enhanced by an ability to 
+redact the transactions from a block prior to its dispatch based on the intent derived from the `SeekInfo` message as done in SmartBFT.
+
+This RFC suggests to build a re-usable component for block pulling that will be used by both the peer and the orderer, and it will have a simple API as the [BlockPuller](https://github.com/hyperledger/fabric/blob/main/orderer/common/cluster/deliver.go#L30-L55):
+
+```
+// BlockFetcher is used to pull blocks from ordering service nodes
+type BlockFetcher interface {
+	PullBlock(seq uint64) *common.Block
+	Close()
+}
+```
+
+The `BlockFetcher` component will perform all bookkeeping required for both CFT and BFT use cases and will then be incorporated in the orderer and peer block replication code.
+This will let the project get rid of multiple implementations that perform the same thing, as currently there is a duplicity in the form of the peer deliver client and the orderer's cluster infrastructure.
+
```

Comment:
Not sure if we need to considering changing SDK/peer endorsement service?

for ex
when any signer submit a Tx.
if the signer x509 information been "cached/stored" in Peer.
return `signer_id(singer identifier information)` in `SignedProposal`
else
keep return x509 information?

as a part of singer identifier information management?

Reply:
This is out of the scope of the RFC.